### PR TITLE
strformat: detect format string errors at compile-time

### PR DESCRIPTION
**Pull request by arnetheduck**
This also prevents unwanted `raises: [ValueError]` effects from bubbling up from correct format strings which makes `fmt` broadly unusable with `raises`.

The old runtime-based `formatValue` overloads are kept for backwards-compatibility, should anyone be using runtime format strings.

### CODE REVIEW

File: lib/pure/strformat.nim [hunk 1]
```diff
@@ -481,50 +481,48 @@ proc parseStandardFormatSpecifier*(s: string; start = 0;
     raise newException(ValueError,
       "invalid format string, cannot parse: " & s[i..^1])
 
+proc toRadix(typ: char): int =
+  case typ
+  of ['x', 'X']: 16
+  of ['d', '\0']: 10
+  of ['o']: 8
+  of ['b']: 2
+  else:
```
Comment (Araq) [inline lib/pure/strformat.nim]:
```suggestion
  of 'x', 'X': 16
  of 'd', '\0': 10
  of 'o': 8
  of 'b': 2
  else:
```

File: lib/pure/strformat.nim [hunk 2]
```diff
@@ -559,23 +557,83 @@ proc formatValue*(result: var string; value: SomeFloat; specifier: string) =
   else:
     result.add res
 
+proc toFloatFormatMode(typ: char): FloatFormatMode =
+  case typ
+  of ['e', 'E']: ffScientific
+  of ['f', 'F']: ffDecimal
+  of ['g', 'G']: ffDefault
+  of ['\0']: ffDefault
```
Comment (Araq) [inline lib/pure/strformat.nim]:
```suggestion
  of 'e', 'E': ffScientific
  of 'f', 'F': ffDecimal
  of 'g', 'G': ffDefault
  of '\0': ffDefault
```

File: lib/pure/strformat.nim [hunk 3]
```diff
@@ -559,23 +557,83 @@ proc formatValue*(result: var string; value: SomeFloat; specifier: string) =
...
+
+    formatFloat(result, value, fmode, spec)
+
+proc formatValue*(result: var string; value: string; specifier: static string) =
+  ## Standard format implementation for `string`. It makes little
+  ## sense to call this directly, but it is required to exist
+  ## by the `&` macro.
+  const spec = parseStandardFormatSpecifier(specifier)
+  var value =
+    when spec.typ in ['s', '\0']: value
```
Comment (Araq) [inline lib/pure/strformat.nim]:
```suggestion
    when spec.typ in {'s', '\0'}: value
```

File: lib/pure/strformat.nim [hunk 4]
```diff
@@ -559,23 +557,83 @@ proc formatValue*(result: var string; value: SomeFloat; specifier: string) =
...
   let spec = parseStandardFormatSpecifier(specifier)
-  var value = value
-  case spec.typ
-  of 's', '\0': discard
-  else:
-    raise newException(ValueError,
-      "invalid type in format string for string, expected 's', but got " &
-      spec.typ)
+  var value =
+    if spec.typ in ['s', '\0']: value
```
Comment (Araq) [inline lib/pure/strformat.nim]:
```suggestion
    if spec.typ in {'s', '\0'}: value
```

File: lib/pure/strformat.nim [hunk 5]
```diff
@@ -559,23 +557,83 @@ proc formatValue*(result: var string; value: SomeFloat; specifier: string) =
...
+    else:
+      raise newException(ValueError,
+        "invalid type in format string for string, expected 's', but got " &
+        spec.typ)
   if spec.precision != -1:
     if spec.precision < runeLen(value):
-      setLen(value, runeOffset(value, spec.precision))
+      let precision = cast[Natural](spec.precision)
+      setLen(value, Natural(runeOffset(value, precision)))
```
Comment (Araq) [inline lib/pure/strformat.nim:629]: Why the change?

Comment (arnetheduck) [reply lib/pure/strformat.nim:629]: compile-time was complaining about converting `int` to `Natural` (in general, I think functions like `runeOffset` that take a `Natural` as input are bugs to be rewritten to `int`, but here we are)

Comment (arnetheduck) [reply lib/pure/strformat.nim:629]: (ie in the runtime version, it's not needed but it felt easier to keep the two the same except for the const vs let)

Comment (Araq) [reply lib/pure/strformat.nim:629]: Why not use a type conversion intead of a `cast` for it?

Comment (arnetheduck) [reply lib/pure/strformat.nim:629]: because I got a compiler error when trying that ;)
